compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPUPackSharedMemoryAlloc.cpp

```diff
@@ -0,0 +1,217 @@
+      // Merge the alloc into the first alias group it interfers with.
+      AllocGroup &mergeGroup = groups[aliasGroups[0]];
+      mergeGroup.allocs.push_back(alloc);
+      Liveness::OperationListT liveInfo =
+          liveness.resolveLiveness(alloc->getResult(0));
+      mergeGroup.liveness.insert(liveInfo.begin(), liveInfo.end());
+      // Then merge all the other alias groups into the first group.
+      for (size_t i = 1, e = aliasGroups.size(); i < e; i++) {
+        AllocGroup &group = groups[aliasGroups[i]];
+        mergeGroup.allocs.insert(mergeGroup.allocs.end(), group.allocs.begin(),
+                                 group.allocs.end());
+        mergeGroup.liveness.insert(group.liveness.begin(),
+                                   group.liveness.end());
+        // For simplicity we leave the group empty and don't remove it.
+        group.allocs.clear();
+        group.liveness.clear();
+      }
+    }
+  }
+
+  LLVM_DEBUG({
+    for (size_t i = 0; i < groups.size(); i++) {
+      llvm::dbgs() << "Alias group " << i << ":\n";
+      for (Operation *op : groups[i].allocs) op->dump();
+    }
+  });
+
+  for (size_t i = 0; i < groups.size(); i++) {
+    if (groups[i].allocs.empty()) continue;
+    aliasGroups.push_back(std::move(groups[i].allocs));
+  }
+}
+
+/// Insert barriers and wait operations if there are allocs of a different alias
+/// group before the given alloc.
+static void addBarrier(func::FuncOp funcOp, Operation *alloc) {
+  Block *entryBlock = &(*funcOp.getBlocks().begin());
+  bool needBarrier = false;
+  if (alloc->getBlock() != entryBlock) {
+    needBarrier = true;
+  } else {
+    for (Operation &op : entryBlock->getOperations()) {
+      if (&op == alloc) break;
+      if (op.getNumRegions() != 0) {
+        needBarrier = true;
+        break;
+      }
+      if (isa<memref::AllocOp>(&op)) {
+        if (!op.hasAttr(kAllocAliasGroup) ||
+            op.getAttr(kAllocAliasGroup) != alloc->getAttr(kAllocAliasGroup)) {
+          needBarrier = true;
+          break;
+        }
+      }
+    }
+  }
+  if (!needBarrier) return;
+  OpBuilder builder(alloc);
+  // TODO: make it a option if needed.
+  bool hasAsyncCopies = true;
+  if (hasAsyncCopies) {
+    Value groupToken = builder.create<nvgpu::DeviceAsyncCreateGroupOp>(
+        funcOp.getLoc(), nvgpu::DeviceAsyncTokenType::get(funcOp.getContext()),
+        SmallVector<Value>());
+    builder.create<nvgpu::DeviceAsyncWaitOp>(funcOp.getLoc(), groupToken,
+                                             builder.getI32IntegerAttr(0));
+  }
+  builder.create<gpu::BarrierOp>(alloc->getLoc());
+}
+
+namespace {
+
+struct LLVMGPUPackSharedMemoryAllocPass
+    : public LLVMGPUPackSharedMemoryAllocBase<
+          LLVMGPUPackSharedMemoryAllocPass> {
+ public:
+  void getDependentDialects(DialectRegistry &registry) const override {
+    registry.insert<nvgpu::NVGPUDialect>();
+  }
+
+  void runOnOperation() override {
+    func::FuncOp funcOp = getOperation();
```

Comment:
Maybe better to move the core functionality into a utility class and make all the gpu specific something that is "configurable by the caller"?

Reply:
I moved `sinkOpsInCFG` into common utils file. The only other part that may be re-used would be  `analyzeSharedMemoryAlloc`, do you have a suggestion on where to put this help? It is a bit opinionated so not sure how to expose it to all backends at the time. I don't see much more to share. What do you think?